### ISSUE TITLE
Support scalar and multidimensional contiguous CPU tensors end to end

### ISSUE THREAD

**Issue by isaacperez** [isaacperez/tabgrad#82]
## Problem

At `main` commit `1d8388e95e0f05c6f30205a6a7b2ee3835f984e5`, the production Python/JavaScript-to-CPU path only admits contiguous rank-one float32 tensors. This restriction appears in runtime shape validation and comparison, `ProgramValue.shape`, CPU element/byte counts, Python input normalization and observation. Adding operations independently would repeat assumptions that a tensor's first dimension is its total size.

## Why it matters

Scalar and multidimensional values are prerequisites for useful tensor calculations. A shared, correctly bounded shape contract is needed by later views, broadcasting, reductions, saved derivative values and serialization. Extending an existing operation through the full path tests that contract without adding an unrelated numerical algorithm or a speculative catalogue of abstractions.

## Expected result

Python and JavaScript users can create contiguous CPU float32 tensors with scalar, vector and higher-rank shapes, add tensors of exactly equal shape through the existing deferred operation, observe the result through the existing observation path, and release resources correctly. Empty dimensions retain their meaning. Shape and size rules have one semantic owner; frontend syntax and physical buffer representation do not redefine them.

This is one bounded production increment, not full tensor-layout support. It extends the accepted semantic/admission and CPU contracts rather than choosing a new runtime architecture.

## Included and excluded work

Included:

- Python `torch.tensor` accepts built-in real scalars and rectangular nested built-in lists/tuples of the already accepted numeric element types, with explicit `torch.float32` and CPU restrictions unchanged. Creation owns a copy. Nested inputs determine only dimensions actually represented by their structure; they do not invent dimensions after an empty sequence.
- JavaScript retains flat iterable/array-like numerical input. The existing `options.shape` accepts scalar `[]` and multidimensional non-negative safe-integer dimensions when the element count matches the input. Omitting shape still produces `[data.length]`. No nested JavaScript data API is introduced.
- Common immutable shape metadata, validated element/byte counts, same-shape admission, executable-program metadata, CPU allocation/kernel lengths and observation use the full shape. The existing coarse scalar/SIMD float32 addition remains the numerical path.
- Ordinary managed Python `tolist()` returns a Python float for rank zero and correctly nested independent lists otherwise. JavaScript `toArray()` remains an independent flat `Float32Array`; `.shape` carries its structure. No JSPI requirement or `_async()` Python method is introduced.
- Existing lifetime, lazy-demand, intermediate-reuse, structured failure and managed-entry guarantees remain applicable to all admitted shapes.

Inspection starts at `src/runtime.ts`, `src/executable-program.ts`, program formation's direct consumers, `src/cpu-backend.ts`, `src/python-runtime-bridge.ts`, `python/torch/__init__.py`, the typed bridge declaration and their direct tests. Inspect Rust/ABI owners only where the established element-count contract or a failure requires it. Start documentation from the Python/JavaScript references and their linked tensor/bridge/program/CPU components; expand only for an affected contract.

Excluded: broadcasting (including scalar-tensor/vector expansion), numerical-scalar operands to addition, new operations, `item()`, additional constructors such as `zeros`/`empty`, views/reshape/transpose/indexing, non-contiguous storage, aliasing/mutation, autograd, new types/promotion/default inference, GPU/transfers, model/weight loading, compilation/fusion, new dependencies, a general resource-policy redesign and package publication. Necessary discoveries follow normal issue handling rather than silently expanding this result.

## Completion conditions

- [ ] Scalar `[]` has one element, `[1]` remains a distinct one-element vector, and any shape containing zero has no numerical elements. Rank-two and rank-three examples preserve all dimensions through creation, pending addition, program formation and observation. Mutating the caller's shape or source containers does not change an admitted tensor.
- [ ] The existing functional, method and operator Python addition forms and JavaScript `.add()` agree on equal full shapes. Two differently shaped tensors with the same element count reject synchronously, including scalar versus `[1]` and different zero-element shapes; no broadcasting or eager numerical work is introduced.
- [ ] Python scalar, rectangular nested list/tuple, singleton dimensions, empty outer input and nested empty inputs produce the documented metadata and result structure. Ragged/mixed-depth inputs, cycles and invalid leaves fail without unbounded traversal, leaked imports or partially admitted tensors. Reusing the same finite subcontainer in separate positions is not mistaken for a cycle.
- [ ] JavaScript shape validation checks every dimension, including dimensions after a zero. Negative, fractional, non-finite, unsafe-integer and non-array shapes reject; product/byte arithmetic cannot wrap or lose integer precision, and non-empty allocations respect the existing address/memory bounds. Oversized or malformed cases use metadata or bounded fixtures, not giant real allocations. A zero element count does not excuse invalid dimensions.
- [ ] Python `tolist()` is normal synchronous-looking code in the established managed entry: scalar results are floats, multidimensional results are nested lists, and successive observations own independent returned data. JavaScript readback stays flat and independently owned. Zero-sized shapes keep full runtime metadata even when an empty returned list cannot encode trailing dimensions.
- [ ] Both CPU scalar and SIMD paths produce oracle-consistent results for the claimed Python subset. Tests cover host-ready, computed and reused-resident values; small higher-rank/empty cases; closed and foreign-session handles; temporary inputs and shared demands; and cleanup after success or a relevant failed import/observation. Existing rank-one, iterative graph formation/release and intermediate-reuse evidence remains valid.
- [ ] Rank generalization adds no per-element bridge/ABI traffic, numerical JavaScript loop, intermediate host round trip or graph-history retention. Shape validation/admission depends on rank/arity, not tensor element count. Payload import/readback and Python container traversal are accounted for separately; nested list overhead is not falsely treated as zero merely because a tensor has no numerical elements.
- [ ] Bounded evidence compares the existing rank-one path against the base and compares equal-element-count contiguous shapes within the new implementation. Record admission/import/observation timing separately, bridge/kernel calls, payload copies, live/peak owned storage and repeated cleanup. Explain variance and material regressions; do not trade them away without an explicit decision. Preserve the existing ownership-safe reuse behavior.
- [ ] References and typed public contracts explain the supported shape/input/output rules and intentional limits. Shared shape concepts and component responsibilities have the appropriate durable home, with operation-specific examples in their references; generic architecture documents do not become an implementation diary. Independent design review verifies that future consumers can reuse shape/size validation without introducing speculative view/autograd machinery.

Develop production behavior with TDD under `docs/quality.md`; use existing oracle generation and real Pyodide/Chrome/Firefox test paths where applicable. No numerical test or benchmark has been run merely by defining this issue.

Measurements run sequentially with one active browser/workload, no worker pool or all-core stress, at most 1,048,576 numerical elements per tensor, at most 8 MiB of experiment-owned numerical payloads and 64 MiB of Tabgrad-owned CPU linear memory, with bounded repetitions and watchdogs. Interpreter/browser memory is separate from that tensor budget. Stop on unexpected growth or resource pressure. High-rank and overflow cases use tiny payloads; no large-model test is required. Record reproducible evidence in the issue/PR, not transient benchmark output in the library tree.

## Dependencies and related work

- #79 is the planning index: this selects a bounded part of C01/H1 and exercises existing C03, C08, C10, C11 and C22/C23. It does not complete those families, start C04/C07, or change their priorities. It is a contextual link, not an executable parent.
- #33 / PR #36 delivered the rank-one JavaScript/CPU path. #43 and #44 / PR #61 established and implemented the bounded Python path. They are completed predecessors whose explicitly excluded rank expansion becomes this new result, not issues to reopen.
- #65 / PR #68, #66 / PR #72, #71 / PR #75 and #76 / PR #78 provide integrated reclamation, iterative traversal and CPU reuse invariants to preserve. They are not outstanding blockers.
- #80 / PR #81 supplies the integrated planning-impact handoff. On completion, reconcile only the actual shape coverage demonstrated here into #79; do not mark general tensor geometry complete.

The duplicate search enumerated all 81 open/closed issue and PR records, compared titles and body matches for multidimensional/rank-aware/rank-zero/rank-one/higher-rank/one-dimensional/scalar-tensor/empty-dimension/tensor-geometry terms, and inspected the relevant predecessor scope and completion evidence. The only open issue is the route index and no open PR competes with this work. There is no existing issue for this production rank extension, no unresolved prerequisite, and no native parent or blocking relationship to add. Existing architecture is the design source; this is not a repeat of historical feasibility experiments.

An independent read-only issue-boundary check by `rank_issue_boundary` examined the original #33/#43/#44 scopes, relevant closing and merged-delivery evidence, the C01/H1 route context, the named interfaces and this complete proposal. It confirmed the distinct follow-on result, absence of an executable parent or open blocker, and no unresolved material public-result choice. The final check completed on 2026-09-23 after an interrupted initial review; it is issue-readiness evidence, not implementation verification or an executed compatibility/performance result.

## Effects and risks

The governing invariant is that logical shape, total element count and physical contiguous storage are related but distinct: identical element counts do not imply identical shapes, scalar rank is not vector rank, and empty dimensions do not erase metadata. Runtime validation owns semantic facts; Python owns normalization of its container syntax; the backend owns physical allocation bounds. No physical pointer or payload is added to semantic program metadata.

The selected boundary keeps contiguous float32 execution and the current operation so shape generality is tested without arbitrary-stride iteration or derivative state. A Python-only matrix special case would duplicate semantics; adding broadcasting/views/types together would obscure which contract failed. Neither is needed for this increment. Exact helper/module layout remains an implementation choice subject to the project's modularity review, not a preselected class catalogue.

Risks include confusing rank with size, overflow before a later zero dimension, ragged/cyclic or deeply nested inputs, rank-proportional stack failures, losing trailing empty dimensions, retaining borrowed buffers, repeated shape walks inside numerical loops, and extra data copies. Preserve structured existing error ownership; distinguish Tabgrad limits from independently measured PyTorch error behavior. Broad support, unlimited rank/allocation and full PyTorch compatibility are not claimed. Resource or representation failures must be explicit, not truncation or fallback.

This extends public accepted inputs and Python result typing while preserving existing rank-one calls, JavaScript flat readback and managed observation requirements. It introduces no new browser requirement, network data flow, runtime dependency or license. Document numerical conversion/input limitations and any representability bounds actually enforced; a material new public limit or architecture change discovered during implementation requires a decision before proceeding.

## User or developer need

A caller needs ordinary scalar and multidimensional calculations through the browser library, while later operations need a reusable shape/size contract instead of repeating vector-only assumptions.

## Example of the intended behavior

Inside the existing managed Python entry:

```python
import torch

left = torch.tensor([[1, 2], [3, 4]], dtype=torch.float32)
right = torch.tensor([[10, 20], [30, 40]], dtype=torch.float32)
result = left + right
assert result.shape == (2, 2)
assert result.tolist() == [[11.0, 22.0], [33.0, 44.0]]

scalar = torch.tensor(2, dtype=torch.float32)
assert scalar.shape == ()
assert (scalar + scalar).tolist() == 4.0

empty = torch.tensor([[], []], dtype=torch.float32)
assert empty.shape == (2, 0)
assert (empty + empty).tolist() == [[], []]
```

JavaScript uses the established flat data interface:

```javascript
const left = session.tensor([1, 2, 3, 4], { shape: [2, 2] });
const right = session.tensor([10, 20, 30, 40], { shape: [2, 2] });
const result = left.add(right);
// result.shape is [2, 2]; Array.from(await result.toArray()) is [11, 22, 33, 44].
```

Existing session/tensor cleanup remains required. These snippets describe intended behavior, not already supported execution.

## Affected public interface

Python `torch.tensor`, `Tensor.shape`, `torch.add` / `Tensor.add` / tensor `+`, and `Tensor.tolist`, plus their checked bridge contracts. JavaScript `RuntimeSession.tensor`'s existing shape option, `Tensor.shape`, `.add()` and `.toArray()`. No new public observation method or constructor is required.

## PyTorch compatibility reference

Use the repository's pinned native PyTorch oracle and fixture provenance for exact expected values, shapes, scalar/list result types and selected errors. Official references inspected while defining this issue:

- [PyTorch 2.14 torch.tensor](https://docs.pytorch.org/docs/2.14/generated/torch.tensor.html): scalar and nested input creation with owned data.
- [PyTorch 2.14 Tensor.tolist](https://docs.pytorch.org/docs/2.14/generated/torch.Tensor.tolist.html): nested lists and a Python number for scalar tensors.
- [PyTorch 2.14 torch.add](https://docs.pytorch.org/docs/2.14/generated/torch.add.html): addition semantics; this issue intentionally retains the narrower equal-shape, CPU float32, alpha=1, out=None contract rather than claiming PyTorch's broadcasting/promotion surface.

Reference documentation is not executed compatibility evidence. JavaScript's flat-array interface remains Tabgrad's own contract. Update the affected reference/support evidence with the production change, without inventing a release-level support claim.

## Deliberately deferred behavior

Views and shared storage, broadcasting/reductions and other operation families, differentiation, randomness, GPU, further types and model/data interoperability remain in #79. They need independently scoped future results, not empty placeholder issues attached to this change.
